### PR TITLE
Feature/faster history getitem

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Changes implementation of `History.__getitem__` so that the type of
items is only looked up once, not each time for each element. This
should be fine, as the types should not change between
elements (e.g. the rows in `'batches'`).

This can save a lot of time. In an example using 1000 samples, 500
epochs and a very small net, history access time goes down from 10 sec
to 2.5 sec.

Here is the code, provided by @stsievert in #719:

```python
import numpy as np
from sklearn.datasets import make_classification
from torch import nn
from skorch import NeuralNetClassifier

class MyModule(nn.Module):
    def __init__(self, n_targets=85):
        super(MyModule, self).__init__()

        self.lin1 = nn.Linear(3, n_targets)
        self.lin2 = nn.Linear(n_targets, 2)
        self.softmax = nn.Softmax(dim=-1)

    def forward(self, X, **kwargs):
        X = self.softmax(self.lin1(X))
        X = self.softmax(self.lin2(X))
        return X

X, y = make_classification(n_samples=1000, n_features=3, n_redundant=1, random_state=0)
X = X.astype(np.float32)
y = y.astype(np.int64)

net = NeuralNetClassifier(
    module=MyModule,
    max_epochs=500,
    lr=0.1,
    iterator_train__shuffle=True,
    verbose=0,
    train_split=False)

net.fit(X, y)
```

### CODE REVIEW

**Comment by stsievert** [comment]
Would it be easier to disable Skorch history recording? In my use case, I'm only concerned about performance and don't care at all about history/callbacks/etc.  I'd be okay with this solution:

``` python
est = NeuralNetClassifier(..., history=False)  # disables recording of history
```

At first, I relied on Skorch for entire training process before realizing that Skorch added a lot of overhead, specifically with callbacks and data conversion from NumPy arrays to PyTorch datasets.

[s]:https://jiffyclub.github.io/snakeviz/
[d]:https://distributed.dask.org/en/latest/diagnosing-performance.html#statistical-profiling

This motivated only using Skorch for initialization and writing a new class that overrode `NeuralNet.partial_fit` and called `torch.from_numpy` before feeding through Skorch's `module_`. I encapsulate this implementation of `partial_fit` in a class `NoOverhead`, which expects NumPy arrays to `NoOverhead.partial_fit`:

``` python
class NoOverhead(NeuralNetRegressor):
    def __init__(self, *args, random_state=None, **kwargs):
        self.random_state = random_state
        super().__init__(*args, **kwargs)

    def partial_fit(self, X: np.ndarray, y: np.ndarray):
        if not self.initialized_:
            self.initialize()
        seed = int(time() * 1000) % (2 ** 30)
        rng = check_random_state(seed)
        X2 = torch.from_numpy(shuffle(X, seed))
        y2 = torch.from_numpy(shuffle(y, seed))

        for k in range(len(X) // self.batch_size):
            lower, upper = k * self.batch_size, (k + 1) * self.batch_size % len(X2)
            y_hat = self.module_.forward(X2[lower:upper])

            loss = self.criterion_(y2[lower:upper], y_hat)

            self.optimizer_.zero_grad()
            loss.backward()
            self.optimizer_.step()
            self.optimizer_.zero_grad()
        return self
```

The complete benchmarking script is a modification of https://github.com/skorch-dev/skorch/pull/730#issue-543098550, and uses similar datasets/models/etc. I use [Dask's statistical profiling tool][d] to profile the timing and see which functions are taking significant time.

<details>
<summary>Complete benchmarking script:</summary>

``` python
import numpy as np
from sklearn.datasets import make_regression
from torch import nn
from skorch import NeuralNetClassifier
from dask.distributed import Client
from skorch import NeuralNetClassifier, NeuralNet, NeuralNetRegressor
from sklearn.utils import check_random_state
import numpy as np
import torch
from time import time

class MyModule(nn.Module):
    def __init__(self, n_targets=85):
        super(MyModule, self).__init__()

        self.lin1 = nn.Linear(3, n_targets)
        self.lin2 = nn.Linear(n_targets, 1)
        self.softmax = nn.Softmax(dim=-1)

    def forward(self, X, **kwargs):
        X = self.softmax(self.lin1(X))
        X = self.softmax(self.lin2(X))
        return X


def shuffle(X: np.ndarray, seed: int) -> np.ndarray:
    Y = X.copy()
    np.random.RandomState(seed).shuffle(Y)
    return Y

class NoOverhead(NeuralNetRegressor):
    def __init__(self, *args, random_state=None, **kwargs):
        self.random_state = random_state
        super().__init__(*args, **kwargs)

    def partial_fit(self, X: np.ndarray, y: np.ndarray):
        if not self.initialized_:
            self.initialize()
        seed = int(time() * 1000) % (2 ** 30)
        rng = check_random_state(seed)
        X2 = torch.from_numpy(shuffle(X, seed))
        y2 = torch.from_numpy(shuffle(y, seed))

        for k in range(len(X) // self.batch_size):
            lower, upper = k * self.batch_size, (k + 1) * self.batch_size % len(X2)
            y_hat = self.module_.forward(X2[lower:upper])

            loss = self.criterion_(y2[lower:upper], y_hat)

            self.optimizer_.zero_grad()
            loss.backward()
            self.optimizer_.step()
            self.optimizer_.zero_grad()
        return self

def main():
    X, y = make_regression(n_samples=1000, n_features=3, random_state=0)
    X = X.astype(np.float32)
    y = y.astype(np.float32).reshape(-1, 1)

    #  est = NoOverhead(
    est = NeuralNetRegressor(
        module=MyModule,
        max_epochs=100,
        criterion=nn.MSELoss,
        batch_size=128,
        lr=0.1,
        iterator_train__shuffle=False,
        train_split=False)
    est.set_params(verbose=0, callbacks__train_loss=None)
    est.callbacks_ = []
    start = time()
    est.partial_fit(X, y)
    return time() - start

if __name__ == "__main__":
    client = Client("localhost:8786")
    future = client.submit(main)
    t = future.result()
    print(t)
```

</details>

 The results:

* `NeuralNetRegressor`: 1.58s.
* `NoOverhead`: 0.00616s, or 6.16ms.

`NeuralNetRegressor` spend the half (or 52.7%) of the time in PyTorch doing data conversions ([dataloader.py#L521][1]). It spends the other half on the core training in `step_fn`. There must be some other overhead in that method.


[1]:https://github.com/pytorch/pytorch/blob/c0deb231db76dbea8a9d326401417f7d1ce96ed5/torch/utils/data/dataloader.py#L521
[py-spy]:https://github.com/benfred/py-spy

**Comment by BenjaminBossan** [comment]
Interesting analysis. A minor amendment: Your `NoOverhead` code runs only a single epoch, whereas the vanilla skorch code runs 100 epochs, so the difference is not quite as drastic as it may seem.

I replicated your results and made some minor adjustments (e.g. made the model actually learn something on the data to verify that all approaches end in a similar result). For measuring performance, I used snakeviz (average of 3 runs). I worked on this branch (i.e. fast history access). Here is what I found:

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">

<colgroup>
<col  class="org-left" />

<col  class="org-right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">condition</th>
<th scope="col" class="org-right">timing</th>
</tr>
</thead>

<tbody>
<tr>
<td class="org-left">vanilla</td>
<td class="org-right">3.0</td>
</tr>

<tr>
<td class="org-left">verbose=0</td>
<td class="org-right">2.1</td>
</tr>

<tr>
<td class="org-left">verbose=0, train_loss=None</td>
<td class="org-right">2.0</td>
</tr>

<tr>
<td class="org-left">verbose=0, history disabled</td>
<td class="org-right">1.9</td>
</tr>

<tr>
<td class="org-left">callbacks disabled</td>
<td class="org-right">1.7</td>
</tr>

<tr>
<td class="org-left">callbacks disabled, fit with tensors*</td>
<td class="org-right">1.6</td>
</tr>

<tr>
<td class="org-left">callbacks disabled, fit with TensorDataset*</td>
<td class="org-right">1.1</td>
</tr>

<tr>
<td class="org-left">no overhead net</td>
<td class="org-right">0.3</td>
</tr>
</tbody>
</table>

\*It's possible to fit skorch nets with torch tensors or torch datasets.

So there is still a pretty big difference, although not nearly as big as initially reported.

What's nice to see is that disabling history is not such a big deal anymore with the proposed changes.

When it comes to the normal PyTorch machinery (forward, backward, etc.), it seems that there is no significant difference between the no-overhead vs normal skorch code. The biggest difference comes through the `DataLoader`, as you also observed. 

Your code completely side-steps PyTorch's `Dataset` and `DataLoader` part. We knew right from the start that these parts would add an overhead. `Dataset` only allows single sample access via `__getitem__`. So effectively, we need to iterate over all individual rows, which are later collated again inside `DataLoader`. Your code iterates over batches, saving a lot of computation.

Could we do away with `Dataset` and `DataLoader` in skorch? I don't think so. They are essential parts of the PyTorch ecosystem, many libraries like torchvision make use of `Dataset`. Once you have `Dataset`s, you need something akin to `DataLoader`. Maybe it would be nice if `Dataset.__getitem__` supported passing slices, not just ints, so that single row access was no longer necessary. But we cannot change that. On top of that, `DataLoader` has some nice properties, like batching of "complex" data (say, a dict of tensors), shuffling, multiple workers etc. If we got rid of it, we would lose this functionality (and might have to re-implement it).

skorch was designed from the start to be PyTorch + extra. Obviously, this extra would have to cost something. We made the assumption that the cost of forward pass, gradient computation, and parameter updates would dominate, and those cannot be avoided.

Still, if we can find a way to squeeze out more performance without making skorch too complex, we should do so (hence this PR). E.g. for this benchmark, I could see that `skorch.dataset.Dataset.__getitem__` is twice as slow as `torch.utils.data.TensorDataset.__getitem__`, adding 0.4 sec to the total. Of course, our `Dataset` has to deal with a lot of different data types, but maybe we can still find room for improvement.

Would it be possible to easily sidestep the whole `Dataset` + `DataLoader` machinery for simple cases? I don't think so without introducing massive changes. This is probably part of the extra that people opt into when using skorch compared to pure PyTorch.

What we should do no matter what is adding some documentation about performance overhead. I think a few of the findings in this thread would fit there nicely. As the table above shows, you can already get from 3 sec to 1.1 sec with a few tweaks.

**Comment by stsievert** [comment]
> A minor amendment: Your NoOverhead code runs only a single epoch, whereas the vanilla skorch code runs 100 epochs, so the difference is not quite as drastic as it may seem.

Thanks for that – I knew the results were too drastic. I can reproduce your numbers, and verified that `partial_fit` runs for `self.max_epochs` epochs (which does not mirror [SGDClassifier][1] as expected).

> Would it be possible to easily sidestep the whole Dataset + DataLoader machinery for simple cases? I don't think so without introducing massive changes. This is probably part of the extra that people opt into when using skorch compared to pure PyTorch.

I have some ideas on how to make that happen (having [get_dataset][2] return an iterable, than catching that in [get_iterator][3]). Hopefully won't require too many compromises (I suspect `tran_split is None` will be necessary). I'll try to submit that PR.

[3]:https://github.com/skorch-dev/skorch/blob/7c76e78238e7edd2bd78afd29a16916438bd4f38/skorch/net.py#L1318
[2]:https://github.com/skorch-dev/skorch/blob/7c76e78238e7edd2bd78afd29a16916438bd4f38/skorch/net.py#L1212
[1]:https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html

**Comment by BenjaminBossan** [comment]
> (which does not mirror [SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html) as expected)

That's [another discussion](https://github.com/skorch-dev/skorch/issues/674).

> I have some ideas on how to make that happen

Yes, I think overriding the mentioned methods could make it feasible to sidestep the whole `Dataset` and `DataLoader` issue by returning an iterator. Depending on requirements, it would need to take care of batching, shuffling, sampling etc. I still think it's probably not feasible to provide this as a general skorch option without causing trouble ("why is `iterator_train__foo=bar` not being used by my iterator?").

> I'll try to submit that PR.

Also take a look at [`get_split_dataset`](https://github.com/skorch-dev/skorch/blob/1c8fc61364d04addebdc89ec4597a1f88e3e50db/skorch/net.py#L1263), which is being called in the fit loop.

**Review by ottonemo — APPROVED** [review]

File: CHANGES.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 
 - We no longer pass the `epoch` parameter to LR schedulers, since that parameter has been deprecated. We now rely on the scheduler to keep track of the epoch itself.
+- Changed implementation of `net.history` access to make it faster; this should not have any noticeable effects, but if you encounter bugs, please create an issue
```
Comment (stsievert) [inline CHANGES.md:20]:
```suggestion
- Changed implementation of `net.history` access to make it faster, which should be most noticeable for small models/datasets.
```

File: skorch/history.py [hunk 2]
```diff
@@ -23,17 +23,108 @@ def _filter_none(items):
...
+
+def _getitem_dict_tuple(item, keys):
+    return tuple(item.get(key, _none) for key in keys)
+
+
+def _getitem_dict_str(item, key):
+    return item.get(key, _none)
+
+
+def _get_getitem_method(items, key):
```
Comment (thomasjpfan) [inline skorch/history.py:92]: I think we can take advantage of https://docs.python.org/3/library/functools.html#functools.singledispatch here instead of dispatching ourselves.

File: skorch/history.py [hunk 3]
```diff
@@ -23,17 +23,108 @@ def _filter_none(items):
...
+        row = []
+        do_append = True
+        for key in keys:
+            try:
+                row.append(item[key])
+            except KeyError:
+                do_append = False
+                break
+        if row and do_append:
+            filtered.append(row)
```
Comment (thomasjpfan) [inline skorch/history.py:44]: No strong opinion:

```py
        for key in keys:
            try:
                row.append(item[key])
            except KeyError:
                break
        else:  # no break
        	if row:
        		filtered.append(row)
```

File: skorch/history.py (hunk 2 above)
Comment (BenjaminBossan) [reply skorch/history.py:92]: Yes, I actually used implemented it both ways. After running a profiler, I saw that there was no difference (presumably `singledispatch` does more or less the same work under the hood).

The main reasons why I didn't go for `singledispatch`:

- we technically need multiple dispatch (`key` and `items`)
- it can look a bit daunting (having to add `@fun.register` everywhere, no nice syntax as in, say, Julia)
- it's presumably not so well known, so harder for readers to understand what's happening, whereas currently, this suggestion only uses "basic" Python language features

What would be the advantages of using `singledispatch` in your opinion?

File: skorch/history.py (hunk 3 above)
Comment (BenjaminBossan) [reply skorch/history.py:44]: Yes, absolutely! I will fix it.

File: CHANGES.md (hunk 1 above)
Comment (BenjaminBossan) [reply CHANGES.md]: Yes, thanks, for sure this entry should mention what the benefit is. I included it but didn't use exactly your proposal.

File: skorch/history.py (hunk 2 above)
Comment (stsievert) [reply skorch/history.py:91]: I haven't looked too closely at the implementation, but I think one benefit of Python's singledispatch is that better documentation is available. Additionally, the concise naming will very likely help new developers like me find relevant examples/SO questions.

Comment (BenjaminBossan) [reply skorch/history.py:91]: > one benefit of Python's singledispatch is that better documentation is available

Do you mean the documentation for `singledispatch` itself? Arguably, the current code doesn't even need consulting documentation as long as you know what `if isinstance(...)` does :) But readability is of course debatable.

However, consider that for this specific case, we dispatch on two arguments, so we would need either nested `singledispatch` or a mix of `singledispatch` and `isinstance`. If we were dispatching only on one argument or if there were a `multidispatch`, I would tend to agree that this approach is superior.

Comment (stsievert) [reply skorch/history.py:91]: Yeah, nevermind. I misunderstood what the comment was referring too. Multiple `isinstance` blocks is a lot clearer.

Comment (thomasjpfan) [reply skorch/history.py:91]: Lets leave it as is. The current implementation is already a net benefit.

**Review by thomasjpfan — APPROVED** [review]
